src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -838,7 +841,7 @@ public AbstractStringBuilder append(char c) {
     public AbstractStringBuilder append(int i) {
         int count = this.count;
         int spaceNeeded = count + DecimalDigits.stringSize(i);
-        ensureCapacityInternal(spaceNeeded);
+        byte[] value = ensureCapacityInternal(spaceNeeded);
         if (isLatin1()) {
```

Comment:
This is not safe. The ensureCapacityInternal can read coder == LATIN1 and allocate a small array, but this `isLatin1` can read coder == UTF16 and write a UTF16 number out of bounds.

Reply:
A check that `spaceNeeded < (value.length >> 1)`  in the `else` branch would be needed and might be a sufficient safeguard here.

Reply:
I made further improvements to improve the thread safety of the coder by passing the newCapacity method into the coder. I think this should be safe enough.